test/test-parse.cpp

```diff
@@ -539,6 +540,208 @@ TEST(test_escape_first_character_in_keyword,
   EXPECT_EQ(escape_first_character_in_keyword(u8"b1n z"_sv), u8"\\u{62}1n z");
   EXPECT_EQ(escape_first_character_in_keyword(u8"ZYXW"_sv), u8"\\u{5a}YXW");
 }
+
+TEST(test_no_overflow, parser_depth_limit_not_exceeded) {
+  {
+    for (const std::pair<char8, char8> &paren :
+         {std::make_pair(u8'(', u8')'), std::make_pair(u8'[', u8']'),
+          std::make_pair(u8'{', u8'}')}) {
+      const int rep_count = parser::stack_limit - 2;
+      const string8 literal = u8"10";
+      string8 brackets(rep_count, paren.first);
+      brackets.reserve(rep_count * 2 + literal.size());
+      brackets.append(literal);
+      for (int i = 0; i < rep_count; i++) {
+        brackets.push_back(paren.second);
+      }
+      padded_string code(brackets);
+      spy_visitor v;
+      parser p(&code, &v);
+      bool ok = p.parse_and_visit_module_catching_fatal_parse_errors(v);
+      EXPECT_TRUE(ok);
+      EXPECT_THAT(v.errors, IsEmpty());
+    }
+  }
+
+  {
+    for (const string8 &exp :
+         {u8"if(true) ", u8"while(true) ", u8"for(;;)", u8"await"}) {
+      const int rep_count = parser::stack_limit - 2;
+      const string8 literal = u8"10";
+      string8 exps;
+      exps.reserve(exp.size() * rep_count + literal.size());
+      for (int i = 0; i < rep_count; i++) {
+        exps.append(exp);
+      }
+      exps.append(literal);
+      padded_string code(exps);
+      spy_visitor v;
+      parser p(&code, &v);
+      bool ok = p.parse_and_visit_module_catching_fatal_parse_errors(v);
+      EXPECT_TRUE(ok);
+      EXPECT_THAT(v.errors, IsEmpty());
+    }
+  }
+
+  {
+    for (const string8 &exp :
+         {u8"function f(){ ", u8"() => { ", u8"if(true){ ", u8"while(true){ ",
+          u8"for(;;){ ", u8"with({}){ "}) {
+      const char8 right_square_bracket = u8'}';
+      const int rep_count = parser::stack_limit / 2 - 1;
+      string8 exps;
+      exps.reserve((exp.size() + 1) * rep_count);
+      for (int i = 0; i < rep_count; i++) {
+        exps.append(exp);
+      }
+      for (int i = 0; i < rep_count; i++) {
+        exps.push_back(right_square_bracket);
+      }
+      padded_string code(exps);
```

Comment:
Maybe we could make a function to help build these strings:

```c++
padded_string repeated_code(string8_view before,
                            string8_view inner,
                            string8_view after,
                            int depth);
```

```suggestion
      padded_string code = repeated_code(exp, u8"", u8"}", parser::stack_limit / 2 - 1);
```